File: Top 100 Codes/05-Operations on Strings/12-Remove brackets from an algebraic expression/Removebracketsfromanalgebraicexpression.cpp
#include <iostream>
#include <stack>
using namespace std;

// Function to remove brackets from an algebraic expression
string removeBrackets(string expr) {
    stack<int> signs; // Stack to keep track of signs
    signs.push(1);    // Assume the expression starts with a positive sign
    
    string result = "";
    int sign = 1; // Current sign
    
    for (int i = 0; i < expr.size(); i++) {
        if (expr[i] == '+') {
            result += (sign == 1) ? '+' : '-';
        } 
        else if (expr[i] == '-') {
            result += (sign == 1) ? '-' : '+';
        } 
        else if (expr[i] == '(') {
            if (i > 0 && expr[i - 1] == '-') {
                sign *= -1;
            }
            signs.push(sign);
        } 
        else if (expr[i] == ')') {
            sign = signs.top();
            signs.pop();
        } 
        else {
            result += expr[i];
        }
    }
    
    // Remove leading '+' if present
    if (!result.empty() && result[0] == '+') {
        result = result.substr(1);
    }
    
    return result;
}

// Driver code
int main() {
    string expression;
    cout << "Enter an algebraic expression: ";
    cin >> expression;

    string result = removeBrackets(expression);
    cout << "Expression without brackets: " << result << endl;

    return 0;
}

File: Top 100 Codes/05-Operations on Strings/12-Remove brackets from an algebraic expression/Removebracketsfromanalgebraicexpression.py
def remove_brackets(expression):
    stack = [1]  # Stack to track signs, assume the first sign is positive
    sign = 1  # Current sign effect
    result = ""

    i = 0
    while i < len(expression):
        if expression[i] == '+':
            result += '+' if sign == 1 else '-'
        elif expression[i] == '-':
            result += '-' if sign == 1 else '+'
        elif expression[i] == '(':
            if i > 0 and expression[i - 1] == '-':
                sign *= -1  # Flip sign inside brackets
            stack.append(sign)
        elif expression[i] == ')':
            sign = stack.pop()  # Restore previous sign
        else:
            result += expression[i]
        i += 1
    
    # Remove leading '+' if present
    if result and result[0] == '+':
        result = result[1:]

    return result

# Example Usage
expression = input("Enter an algebraic expression: ")
print("Expression without brackets:", remove_brackets(expression))
